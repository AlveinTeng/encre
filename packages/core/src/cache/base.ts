import hash from 'object-hash';
import { Generation } from '../events/output/provide/generation.js';

export abstract class BaseCache<T = Generation[]> {
  abstract lookup(keysStr: string[]): Promise<T | null>;

  abstract update(keysStr: string[], value: T): Promise<void>;
}

export function getCacheKey(keysStr: string[]): string {
  return hash(keysStr.join('_'));
}
