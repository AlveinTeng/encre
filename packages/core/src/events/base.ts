import { SerializedFields } from '../load/keymap';
import { Callable, CallableConfigFields, CallableConfig } from '../record/callable';

export interface BaseEventParams extends CallableConfig {
  /**
   * Whether to print out response text.
   */
  verbose?: boolean;
}

export abstract class BaseEvent<
    CallInput,
    CallOutput,
    CallOptions extends CallableConfig = CallableConfig,
  >
  extends Callable<CallInput, CallOutput, CallOptions>
  implements BaseEventParams {
  
  /**
   * Whether to print out response text.
   */
  verbose: boolean;

  name?: string;

  tags?: string[];

  metadata?: CallableConfigFields;

  callbacks?: any;

  get _attributes(): SerializedFields | undefined {
    return {
      verbose: undefined,
      callbacks: undefined
    };
  }

  constructor(fields: BaseEventParams) {
    super(fields);
    this.verbose = fields.verbose ?? false;
    this.tags = fields.tags ?? [];
    this.metadata = fields.metadata ?? {};
    this.callbacks = fields.callbacks;
  }
}
