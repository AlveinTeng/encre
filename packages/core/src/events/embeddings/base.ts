import {
    AsyncCallError,
    AsyncCaller,
    type AsyncCallerParams,
  } from '../../utils/asyncCaller.js';

export type EmbeddingsParams = AsyncCallerParams;

export abstract class Embeddings {
    /**
     * The async caller should be used by subclasses to make any async calls,
     * which will thus benefit from the concurrency and retry logic.
     */
    caller: AsyncCaller;
  
    constructor(params: EmbeddingsParams) {
      this.caller = new AsyncCaller(params ?? {});
    }
  
    /**
     * An abstract method that takes an array of documents as input and
     * returns a promise that resolves to an array of vectors for each
     * document.
     * @param documents An array of documents to be embedded.
     * @returns A promise that resolves to an array of vectors for each document.
     */
    abstract embedDocuments(documents: string[]): Promise<number[][]>;
  
    /**
     * An abstract method that takes a single document as input and returns a
     * promise that resolves to a vector for the query document.
     * @param document A single document to be embedded.
     * @returns A promise that resolves to a vector for the query document.
     */
    abstract embedQuery(document: string): Promise<number[]>;
  }