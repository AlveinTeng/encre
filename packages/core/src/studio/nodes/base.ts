import {
  RecordId,
  SecretFields,
  SerializedFields,
  SerializedKeyAlias,
} from '../../load/keymap.js';
import { SerializedCallableFields } from '../../record/callable.js';
import { isNotNull } from '../../utils/safeTypes.js';
import { DataFields, DataType } from '../data.js';
import { NodeProcessInfo } from '../graph.js';
import {
  ProcessContext,
  ProcessInputMap,
  ProcessOutputMap,
  validateProcessDataFromPorts,
} from '../processor.js';
import { GuardrailRegistration } from '../registration/guardrails.js';
import { NodeRegistration } from '../registration/nodes.js';
import { SerializedNode } from '../serde.js';
import { UIContext } from '../ui.js';
import { coerceToData } from '../utils/coerce.js';
import {
  displayUIFromDataFields,
  displayUIFromSecretFields,
} from '../utils/display.js';
import {
  CallableNode,
  NodeBody,
  NodeConnection,
  NodeInputPortDef,
  NodeOutputPortDef,
  NodePortDef,
  NodePortFields,
  NodePortSizes,
  SerializableNode,
} from './index.js';

export abstract class NodeImpl<
  T extends SerializableNode,
  Type extends T['type'] = T['type'],
  SubType extends T['subType'] = T['subType'],
> {
  readonly node: T;

  readonly aliases: SerializedKeyAlias;

  readonly secrets: SecretFields;

  private _kwargs: DataFields;

  constructor(node: T) {
    this.node = node;

    const { aliases, secrets, kwargs } = this.node.data.getAttributes();
    this.aliases = aliases;
    this.secrets = secrets;
    this._kwargs = this._coerceSerializedToData(kwargs);
  }

  get title(): string {
    return this.node.title ?? this.name;
  }

  get description(): string {
    return this.node.description ?? '';
  }

  get id(): RecordId {
    return this.node.id;
  }

  get type(): Type {
    return this.node.type as Type;
  }

  get subType(): SubType {
    return this.node.subType as SubType;
  }

  get registerArgs(): Record<string, unknown> | undefined {
    return this.node.registerArgs;
  }

  get name(): string {
    return this.node.data._id[this.node.data._id.length - 1];
  }

  get visualInfo(): T['visualInfo'] {
    return this.node.visualInfo;
  }

  get data(): T['data'] {
    return this.node.data;
  }

  get inputs(): T['inputs'] {
    return this.node.inputs;
  }

  get outputs(): T['outputs'] {
    return this.node.outputs;
  }

  get outputSizes(): NodePortSizes {
    return this.node.outputSizes ?? this._initSize();
  }

  get kwargs(): DataFields {
    return this._kwargs;
  }

  get runtime(): number {
    return this.node.runtime ?? 0;
  }

  get memory(): number {
    return this.node.memory ?? 0;
  }

  setKwarg(key: string, value: unknown): void {
    if (!(key in this._kwargs)) {
      throw new Error(`keyword ${key} does not exist in ${this.title}`);
    }

    this.node.data[key] = value;

    const { kwargs } = this.node.data.getAttributes();
    this._kwargs = this._coerceSerializedToData(kwargs);
  }

  private _coerceSerializedToData(kwargs: SerializedFields): DataFields {
    return Object.fromEntries(
      Object.entries(kwargs).map(([k, v]) => [k, coerceToData(v)])
    );
  }

  getInputPortDefs(
    connections: NodeConnection[] = [],
    nodeMap: Record<RecordId, SerializableNode> = {}
  ): NodeInputPortDef[] {
    // If there is a IfNode connected to the current node, then the
    // connected corresponding ports can be coerced to optional (i.e. unknown)
    const portNamesCoerceToOptional: string[] = connections
      .filter((conn) => {
        const node = nodeMap[conn.fromNodeId];

        return (
          conn.toNodeId === this.id && isNotNull(node) && node.type === 'if'
        );
      })
      .map((conn) => conn.toPortName);

    return this._initPorts(this.inputs, portNamesCoerceToOptional);
  }

  getOutputPortDefs(
    connections: NodeConnection[] = [],
    nodeMap: Record<RecordId, SerializableNode> = {}
  ): NodeOutputPortDef[] {
    return this._initPorts(this.outputs);
  }

  validateInputs(inputs?: ProcessInputMap): boolean {
    // check if there is no input requirement
    if (!this.inputs || Object.keys(this.inputs).length === 0) return true;

    // check if there is no given input
    if (!inputs || Object.keys(inputs).length === 0) return false;

    return validateProcessDataFromPorts(inputs, this.inputs);
  }

  abstract process(
    inputs: ProcessInputMap,
    context: ProcessContext
  ): Promise<ProcessOutputMap>;

  async getBody(): Promise<NodeBody> {
    const secretUIContexts: UIContext[] = await displayUIFromSecretFields(
      this.secrets
    );
    const kwargsUIContexts: UIContext[] = await displayUIFromDataFields(
      this.kwargs
    );

    return [...secretUIContexts, ...kwargsUIContexts];
  }

  static async deserialize(
    serialized: SerializedNode,
    values: Record<string, unknown> = {},
    registry?: {
      nodes?: NodeRegistration;
      guardrails?: GuardrailRegistration;
    }
  ): Promise<SerializableNode> {
    if (serialized._type !== 'node') {
      throw new Error(
        `CANNOT deserialize this type in node: ${serialized._type}`
      );
    }

    switch (serialized.type) {
      case 'graph': {
        const { GraphNodeImpl } = await import('./utility/graph.node.js');
        return GraphNodeImpl.deserialize(serialized, values, registry);
      }
      case 'if': {
        const { IfNodeImpl } = await import('./utility/if.node.js');
        return IfNodeImpl.deserialize(serialized, values, registry);
      }
      case 'splitter': {
        const { SplitterNodeImpl } = await import('./input/splitter.node.js');
        return SplitterNodeImpl.deserialize(serialized);
      }
      case 'prompt': {
        const { PromptNodeImpl } = await import('./input/prompt.node.js');
        return PromptNodeImpl.deserialize(serialized);
      }
      case 'message': {
        const { MessageNodeImpl } = await import('./input/message.node.js');
        return MessageNodeImpl.deserialize(serialized);
      }
      case 'loader': {
        const { LoaderNodeImpl } = await import('./input/loader.node.js');
        return LoaderNodeImpl.deserialize(serialized);
      }
      case 'llm': {
        const { LLMNodeImpl } = await import('./inference/chat/llm.node.js');
        return LLMNodeImpl.deserialize(serialized);
      }
      case 'chatlm': {
        const { ChatLMNodeImpl } = await import(
          './inference/chat/chatlm.node.js'
        );
        return ChatLMNodeImpl.deserialize(serialized);
      }
      default:
        throw new Error('Plugin node is unsupported for now');
    }
  }

  async serialize(
    connections: NodeConnection[],
    processInfo?: NodeProcessInfo
  ): Promise<SerializedNode> {
    const outgoingConnections: {
      [key in string]: { toNodeId: RecordId; toPortName: string };
    } = {};

    for (const conn of connections) {
      const { fromPortName, toNodeId, toPortName } = conn;
      outgoingConnections[fromPortName] = { toNodeId, toPortName };
    }

    return {
      _type: 'node',
      id: this.id,
      type: this.type,
      subType: this.subType,
      registerArgs: this.registerArgs,
      title: this.title,
      description: this.description,
      runtime: processInfo?.runtime ?? this.runtime,
      memory: processInfo?.memory ?? this.memory,
      data: JSON.parse(JSON.stringify(this.data)),
      visualInfo: this.visualInfo,
      inputs: this.inputs,
      outputs: this.outputs,
      outputSizes: this.outputSizes,
      outgoingConnections,
    };
  }

  private _initSize(): NodePortSizes {
    return Object.fromEntries(
      Object.keys(this.outputs ?? {}).map((key: string) => [key, 0])
    );
  }

  private _initPorts(
    ports: NodePortFields | undefined,
    portNamesCoerceToOptional: string[] = []
  ): NodePortDef[] {
    return Object.entries(ports ?? {}).map(([key, type]) => {
      const types: DataType[] = Array.isArray(type) ? type : [type];
      if (
        portNamesCoerceToOptional.includes(key) &&
        !types.includes('unknown')
      ) {
        types.push('unknown');
      }

      return {
        nodeId: this.id,
        name: key,
        type: types,
        required: !types.includes('unknown'),
      };
    });
  }
}

export abstract class CallableNodeImpl<
  T extends CallableNode,
  Type extends T['type'] = T['type'],
  SubType extends T['subType'] = T['subType'],
> extends NodeImpl<T, Type, SubType> {
  private _metadata: {
    type: string;
    callables?: SerializedCallableFields | undefined;
  };

  constructor(node: T) {
    super(node);

    const { metadata } = this.node.data.getAttributes();
    this._metadata = metadata;
  }

  get callableType(): string {
    return this._metadata.type;
  }

  protected abstract _preprocess(
    inputs: ProcessInputMap,
    context: ProcessContext
  ): Promise<unknown>;

  protected abstract _postprocess(
    rawOutputs: unknown,
    context: ProcessContext
  ): Promise<ProcessOutputMap>;

  abstract invoke<CallInput, CallOutput, CallOptions>(
    input: CallInput,
    options?: Partial<CallOptions>
  ): Promise<CallOutput>;
}
