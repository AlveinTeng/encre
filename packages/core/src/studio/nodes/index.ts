import { RecordId } from '../../load/keymap.js';
import { Serializable } from '../../load/serializable.js';
import { Callable } from '../../record/callable.js';
import { DataType, ValueOf } from '../data.js';
import { UIContext } from '../ui.js';

export type NodePortFields = {
  [key: string]: DataType | Readonly<DataType[]>;
};

export interface BaseNode {
  visualInfo: {
    position: {
      x: number;
      y: number;
      zIndex?: number;
    };
    size: {
      width: number;
      height: number;
    };
  };

  inputs: NodePortFields | undefined;

  outputs: NodePortFields | undefined;
}

export interface SerializableNode<
  NodeType extends string = string,
  NodeData extends Serializable = Serializable,
> extends BaseNode {
  type: NodeType;

  subType: string;

  data: NodeData;
}

export interface CallableNode<
  NodeType extends string = string,
  NodeData extends Callable = Callable,
> extends SerializableNode<NodeType, NodeData> {}

export type NodeInputPortDef = {
  id: RecordId;

  name: string;

  type: DataType | Readonly<DataType[]>;

  data?: ValueOf<DataType> | Readonly<ValueOf<DataType>[]>;

  default?: unknown;
};

export type NodeOutputPortDef = {
  id: RecordId;

  name: string;

  type: DataType | Readonly<DataType[]>;

  data?: ValueOf<DataType> | Readonly<ValueOf<DataType>[]>;

  default?: unknown;
};

export type NodeConnection = {
  inputNodeId: RecordId;

  inputId: RecordId;

  outputNodeId: RecordId;

  outputId: RecordId;
};

export type NodeBody =
  | string
  | UIContext
  | UIContext[]
  | undefined;
