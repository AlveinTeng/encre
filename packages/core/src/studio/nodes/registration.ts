import { NodeImpl, NodeImplConstructor } from './base.js';
import {
  GeminiChatNodeImpl,
  OpenAIChatNodeImpl,
} from './inference/chat/chatlm.node.js';
import { GeminiNodeImpl, OpenAINodeImpl } from './inference/chat/llm.node.js';
import { PDFLoaderNodeImpl } from './input/loader.node.js';
import {
  BotMessageNodeImpl,
  ChatMessageNodeImpl,
  FunctionMessageNodeImpl,
  HumanMessageNodeImpl,
  SystemMessageNodeImpl,
} from './input/message.node.js';
import {
  ChatPromptNodeImpl,
  StringPromptNodeImpl,
} from './input/prompt.node.js';
import {
  LanguageTextSplitterNodeImpl,
  RecursiveTextSplitterNodeImpl,
  TextSplitterNodeImpl,
  TokenTextSplitterNodeImpl,
} from './input/splitter.node.js';
import { SerializableNode } from './index.js';

type ExtractType<T> = T extends `${infer U}-${any}` ? U : never;

export class NodeRegistration<
  NodeTypes extends string = never,
  NodeSubTypes extends string = never,
  Nodes extends SerializableNode = never,
> {
  registeredNodes: Nodes = undefined!;

  registeredNodeTypes: NodeTypes = undefined!;

  registeredNodeTypePairs: { [P in NodeTypes]: NodeSubTypes[] } = undefined!;

  info = {} as {
    [P in `${NodeTypes}-${NodeSubTypes}`]: {
      impl: NodeImplConstructor<Extract<Nodes, { type: ExtractType<P> }>>;
    };
  };

  implsMap = {} as Record<
    string,
    { impl: NodeImplConstructor<SerializableNode> }
  >;

  readonly nodeTypes = new Set<NodeTypes>();

  readonly nodeTypePairs = {} as { [P in NodeTypes]: NodeSubTypes[] };

  get nodeConstructors(): NodeImplConstructor<SerializableNode>[] {
    return Object.values(this.implsMap).map((nodeImpl) => nodeImpl.impl);
  }

  register<T extends SerializableNode>(
    impl: NodeImplConstructor<T>,
    args?: Record<string, unknown>
  ): NodeRegistration<
    NodeTypes | T['type'],
    NodeSubTypes | T['subType'],
    Nodes | T
  > {
    const newRegistration = this as NodeRegistration<
      NodeTypes | T['type'],
      NodeSubTypes | T['subType'],
      Nodes | T
    >;

    const node = impl.create(args);
    const type = node.type as T['type'];
    const subType = node.subType as T['subType'];

    const key = [type, subType].join('-');

    if (newRegistration.info[key]) {
      throw new Error(`Duplicate node type pair: ${key}`);
    }

    newRegistration.info[key] = {
      impl: impl as any,
    };

    newRegistration.implsMap[key] = {
      impl: impl as any,
    };

    newRegistration.nodeTypes.add(type);

    if (newRegistration.nodeTypePairs[type]) {
      newRegistration.nodeTypePairs[type].push(subType);
    } else {
      newRegistration.nodeTypePairs[type] = [subType];
    }

    return newRegistration;
  }

  create<T extends NodeTypes>(
    type: T,
    subType: string,
    args?: Record<string, unknown>
  ): Extract<Nodes, { type: T }> {
    const key = [type, subType].join('-');

    const nodeImpl = this.info[key] as {
      impl: NodeImplConstructor<Extract<Nodes, { type: T }>>;
    };

    if (!nodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    return nodeImpl.impl.create(args);
  }

  createImpl<T extends Nodes>(node: T): NodeImpl<T> {
    const type = node.type as Extract<NodeTypes, T['type']>;
    const subType: string = node.subType;

    const key = [type, subType].join('-');

    const nodeImpl = this.info[key] as {
      impl: NodeImplConstructor<Extract<Nodes, { type: T }>>;
    };

    if (!nodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    const { impl: Impl } = nodeImpl;

    const newNodeImpl = new Impl(node as any) as unknown as NodeImpl<T>;

    if (!newNodeImpl) {
      throw new Error(`Unknown node type pair: ${key}`);
    }

    return newNodeImpl;
  }

  isRegistered(type: string, subType: string): boolean {
    const key = [type, subType].join('-');

    return this.info[key] !== undefined;
  }
}

export function registerBuiltInNodes(registry: NodeRegistration) {
  return registry
    .register(PDFLoaderNodeImpl)
    .register(ChatMessageNodeImpl)
    .register(HumanMessageNodeImpl)
    .register(BotMessageNodeImpl)
    .register(SystemMessageNodeImpl)
    .register(FunctionMessageNodeImpl)
    .register(StringPromptNodeImpl)
    .register(ChatPromptNodeImpl)
    .register(TextSplitterNodeImpl)
    .register(RecursiveTextSplitterNodeImpl)
    .register(LanguageTextSplitterNodeImpl, { language: 'cpp' })
    .register(LanguageTextSplitterNodeImpl, { language: 'go' })
    .register(LanguageTextSplitterNodeImpl, { language: 'java' })
    .register(LanguageTextSplitterNodeImpl, { language: 'js' })
    .register(LanguageTextSplitterNodeImpl, { language: 'php' })
    .register(LanguageTextSplitterNodeImpl, { language: 'proto' })
    .register(LanguageTextSplitterNodeImpl, { language: 'python' })
    .register(LanguageTextSplitterNodeImpl, { language: 'rst' })
    .register(LanguageTextSplitterNodeImpl, { language: 'ruby' })
    .register(LanguageTextSplitterNodeImpl, { language: 'rust' })
    .register(LanguageTextSplitterNodeImpl, { language: 'scala' })
    .register(LanguageTextSplitterNodeImpl, { language: 'markdown' })
    .register(LanguageTextSplitterNodeImpl, { language: 'latex' })
    .register(LanguageTextSplitterNodeImpl, { language: 'html' })
    .register(LanguageTextSplitterNodeImpl, { language: 'sol' })
    .register(TokenTextSplitterNodeImpl)
    .register(OpenAINodeImpl)
    .register(GeminiNodeImpl)
    .register(OpenAIChatNodeImpl)
    .register(GeminiChatNodeImpl);
}

let globalNodeRegistry = registerBuiltInNodes(new NodeRegistration());

export { globalNodeRegistry };

export type BuiltInNodes = typeof globalNodeRegistry.registeredNodes;

export type BuiltInNodeTypes = typeof globalNodeRegistry.registeredNodeTypes;

export type BuiltInNodeTypePairs =
  typeof globalNodeRegistry.registeredNodeTypePairs;

export type NodeOf<T extends BuiltInNodeTypes> = Extract<
  BuiltInNodes,
  { type: T }
>;

export function resetNodeRegistry() {
  globalNodeRegistry = registerBuiltInNodes(new NodeRegistration());
}
