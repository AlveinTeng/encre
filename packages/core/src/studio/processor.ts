import type { Opaque } from 'type-fest';
import { RecordId } from '../load/keymap.js';

import { Data, DataType } from './data.js';

import { NodePortFields } from './nodes/index.js';

export type ProcessId = Opaque<string, 'ProcessId'>;

export type ProcessInputMap = Record<string, Data | undefined>;
export type ProcessOutputMap = Record<string, Data | undefined>;

export type ProcessDataMap = ProcessInputMap | ProcessOutputMap;

export type ProcessContext = {
  processId: ProcessId;

  signal: AbortSignal;

  contextData: Record<string, Data>;

  graphInputs: Record<string, Data>;

  graphOutputs: Record<string, Data>;

  // In the memory, we store callable options for CallableNodes
  memory: Record<RecordId, unknown>;
};

/**
 * Validate whether the ports are full-filled with the process data.
 */
export function validateProcessDataFromPorts(
  processData: ProcessDataMap,
  portFields: NodePortFields
): boolean {
  const keywords: string[] = Object.keys(portFields);

  return keywords.every((keyword: string): boolean => {
    // Return false if there is no such a port keyword in the process data.
    if (!Object.prototype.hasOwnProperty.call(processData, keyword)) {
      return false;
    }

    const processValue: Data | undefined = processData[keyword];
    const portType: DataType | Readonly<DataType[]> = portFields[keyword];

    return validateProcessData(processValue, portType);
  });
}

export function validateProcessData(
  processData: Data | undefined,
  dataType: DataType | Readonly<DataType[]>
): boolean {
  // Check if this keyword can support multiple types of data
  if (Array.isArray(dataType)) {
    // Return true if the process data value and the type are both empty.
    if (!processData) {
      return (
        dataType.length === 0 || dataType.some((t: DataType) => t === 'unknown')
      );
    }

    // Return true if there is some type in the type array that can be valid.
    // TODO: think about if coerceTypeOptional can apply here
    return dataType.some((t: DataType) => processData['type'] === t);
  }

  // Return false if the process data value is empty and the type is non-empty.
  if (!processData) return dataType === 'unknown';

  // Validate the types
  // TODO: think about if coerceTypeOptional can apply here
  return processData['type'] === dataType;
}
