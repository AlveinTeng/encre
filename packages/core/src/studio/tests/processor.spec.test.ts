import { beforeAll, describe, expect, test } from '@jest/globals';
import { SubGraph } from '../graph';
import { CallableNodeImpl } from '../nodes/base';
import { GraphInputs, GraphProcessor, ProcessInputMap } from '../processor';
import { globalNodeRegistry } from '../registration/nodes';

describe('GraphProcessor', () => {
  test('test simple node in graph', async () => {
    const OpenAINode = globalNodeRegistry.createDynamic('chatlm', 'openai');

    const graph = new SubGraph({ nodes: [OpenAINode], connections: [] });

    const processor = new GraphProcessor(graph);

    const processInput: ProcessInputMap = {
      prompt: {
        type: 'string',
        value: 'Who are you?',
      },
    };

    const graphInputs: GraphInputs = {};
    graphInputs[OpenAINode.id] = processInput;

    expect(
      graph.nodeImplMap[OpenAINode.id].getInputPortDefs()
    ).toMatchSnapshot();

    expect(
      await (graph.nodeImplMap[OpenAINode.id] as CallableNodeImpl<any>).invoke(
        'Who are you?'
      )
    ).toMatchSnapshot();

    expect(
      await graph.nodeImplMap[OpenAINode.id].process(processInput, {})
    ).toMatchSnapshot();

    expect(await processor.processGraph(graphInputs)).toMatchSnapshot();
  });
});
