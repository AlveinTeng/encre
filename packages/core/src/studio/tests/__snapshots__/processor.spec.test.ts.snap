// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphProcessor test simple node in graph 1`] = `
[
  {
    "#type": "processInQueue",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "QmwuQb1MeHs2cLin@",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processorId": "UVLKvO@vV7iYpot5v",
  },
  {
    "#type": "nodeStart",
    "inputs": {
      "prompt": {
        "type": "string",
        "value": "Who are you?",
      },
    },
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "QmwuQb1MeHs2cLin@",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "processId": "UUJPV735QPORnfHU7",
    "processorId": "UVLKvO@vV7iYpot5v",
  },
  {
    "#type": "newAbortController",
    "controller": AbortController {},
    "isSubProcessor": false,
    "processorId": "UVLKvO@vV7iYpot5v",
  },
  {
    "#type": "nodeFinish",
    "isSubProcessor": false,
    "node": {
      "data": {
        "_grp": 1,
        "_id": [
          "events",
          "inference",
          "chat",
          "chatlms",
          "openai",
          "OpenAIChat",
        ],
        "_kwargs": {
          "additional_kwargs": {},
          "callbacks": undefined,
          "frequency_penalty": 0,
          "logit_bias": undefined,
          "max_tokens": 2048,
          "model": "gpt-3.5-turbo",
          "openai_api_key": {
            "_grp": 1,
            "_id": [
              "OPENAI_API_KEY",
            ],
            "_type": "secret",
          },
          "presence_penalty": 0,
          "response_format_type": undefined,
          "seed": undefined,
          "stop_words": undefined,
          "stream": false,
          "temperature": 1,
          "timeout": undefined,
          "top_p": 1,
          "user": undefined,
          "verbose": undefined,
        },
        "_type": "constructor",
      },
      "id": "QmwuQb1MeHs2cLin@",
      "inputs": {
        "prompt": [
          "string",
          "chat-message[]",
        ],
      },
      "outputs": {
        "info": [
          "object",
          "unknown",
        ],
        "message": "chat-message",
        "output": "string",
        "tokenUsage": [
          "object",
          "unknown",
        ],
      },
      "subType": "openai",
      "type": "chatlm",
      "visualInfo": {
        "position": {
          "x": 0,
          "y": 0,
        },
        "size": {
          "height": 500,
          "width": 300,
        },
      },
    },
    "outputs": {
      "info": {
        "type": "object",
        "value": {
          "finishReason": "stop",
        },
      },
      "message": {
        "type": "chat-message",
        "value": {
          "_grp": 1,
          "_id": [
            "events",
            "input",
            "load",
            "msgs",
            "assistant",
            "BotMessage",
          ],
          "_kwargs": {
            "additional_kwargs": {},
            "content": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
          },
          "_type": "constructor",
        },
      },
      "output": {
        "type": "string",
        "value": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
      },
      "tokenUsage": {
        "type": "object",
        "value": {
          "completionTokens": 33,
          "promptTokens": 11,
          "totalTokens": 44,
        },
      },
    },
    "processId": "UUJPV735QPORnfHU7",
    "processorId": "UVLKvO@vV7iYpot5v",
  },
  {
    "#type": "graphFinish",
    "graph": {
      "_grp": 1,
      "_id": [
        "studio",
        "graph",
        "SubGraph",
      ],
      "_kwargs": {
        "connections": [],
        "nodes": [
          {
            "data": {
              "_grp": 1,
              "_id": [
                "events",
                "inference",
                "chat",
                "chatlms",
                "openai",
                "OpenAIChat",
              ],
              "_kwargs": {
                "additional_kwargs": {},
                "callbacks": undefined,
                "frequency_penalty": 0,
                "logit_bias": undefined,
                "max_tokens": 2048,
                "model": "gpt-3.5-turbo",
                "openai_api_key": {
                  "_grp": 1,
                  "_id": [
                    "OPENAI_API_KEY",
                  ],
                  "_type": "secret",
                },
                "presence_penalty": 0,
                "response_format_type": undefined,
                "seed": undefined,
                "stop_words": undefined,
                "stream": false,
                "temperature": 1,
                "timeout": undefined,
                "top_p": 1,
                "user": undefined,
                "verbose": undefined,
              },
              "_type": "constructor",
            },
            "id": "QmwuQb1MeHs2cLin@",
            "inputs": {
              "prompt": [
                "string",
                "chat-message[]",
              ],
            },
            "outputs": {
              "info": [
                "object",
                "unknown",
              ],
              "message": "chat-message",
              "output": "string",
              "tokenUsage": [
                "object",
                "unknown",
              ],
            },
            "subType": "openai",
            "type": "chatlm",
            "visualInfo": {
              "position": {
                "x": 0,
                "y": 0,
              },
              "size": {
                "height": 500,
                "width": 300,
              },
            },
          },
        ],
      },
      "_type": "constructor",
    },
    "isSubProcessor": false,
    "outputs": {
      "QmwuQb1MeHs2cLin@": {
        "info": {
          "type": "object",
          "value": {
            "finishReason": "stop",
          },
        },
        "message": {
          "type": "chat-message",
          "value": {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "assistant",
              "BotMessage",
            ],
            "_kwargs": {
              "additional_kwargs": {},
              "content": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
            },
            "_type": "constructor",
          },
        },
        "output": {
          "type": "string",
          "value": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
        },
        "tokenUsage": {
          "type": "object",
          "value": {
            "completionTokens": 33,
            "promptTokens": 11,
            "totalTokens": 44,
          },
        },
      },
    },
    "processorId": "UVLKvO@vV7iYpot5v",
  },
  {
    "#type": "done",
    "isSubProcessor": false,
    "processorId": "UVLKvO@vV7iYpot5v",
    "results": {
      "QmwuQb1MeHs2cLin@": {
        "info": {
          "type": "object",
          "value": {
            "finishReason": "stop",
          },
        },
        "message": {
          "type": "chat-message",
          "value": {
            "_grp": 1,
            "_id": [
              "events",
              "input",
              "load",
              "msgs",
              "assistant",
              "BotMessage",
            ],
            "_kwargs": {
              "additional_kwargs": {},
              "content": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
            },
            "_type": "constructor",
          },
        },
        "output": {
          "type": "string",
          "value": "I am a language model AI created by OpenAI. I am here to assist you with any questions or tasks you may have. How can I help you today?",
        },
        "tokenUsage": {
          "type": "object",
          "value": {
            "completionTokens": 33,
            "promptTokens": 11,
            "totalTokens": 44,
          },
        },
      },
    },
  },
  {
    "#type": "finish",
    "isSubProcessor": false,
    "processorId": "UVLKvO@vV7iYpot5v",
  },
]
`;
